File: 2021/Python/day_03/day_03.py
with open('input', 'r') as f :
    numbers = [line.strip() for line in f.readlines()]

bit_length = len(numbers[0])
number_rows = len(numbers)


############ Part 1

counts = [0] * bit_length

for pattern in numbers :
    for i, c in enumerate(pattern) :
        if c == "1" :
            counts[i] += 1

gamma_rate = ""
epsilon_rate = ""

for d in counts :
    if d > number_rows - d :
        gamma_rate += "1"
        epsilon_rate += "0"
    else :
        gamma_rate += "0"
        epsilon_rate += "1"
    
power_consumption = int(gamma_rate, 2) * int(epsilon_rate, 2)
print(f"Part 1. Power consumption: {power_consumption}")


############ Part 2

def filter(numbers, preferOnes=True) :
    candidates = numbers.copy()

    for i in range(len(candidates[0])) :
        if len(candidates) == 1 :
            return candidates[0]

        oners, zeroers = [], []
        for pattern in candidates :
            if pattern[i] == "1" :
                oners.append(pattern)
            else :
                zeroers.append(pattern)
        
        if preferOnes :
            candidates = oners if len(oners) >= len(zeroers) else zeroers
        else :  
            candidates = zeroers if len(zeroers) <= len(oners) else oners
    
    return candidates[0]

oxygen_generator_rating = filter(numbers, preferOnes=True)
co2_scrubber_rating = filter(numbers, preferOnes=False)

life_support_rating = int(oxygen_generator_rating, 2) * int(co2_scrubber_rating, 2)
print(f"Part 2. Life support rating: {life_support_rating}")

File: 2021/Python/day_03/day_03.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cmath>

using binary = std::vector<int>;
using binaryList = std::vector<binary>;

binaryList read_input()
{
    binaryList list;

    std::string line;
    while( std::getline(std::cin, line) )
    {
        binary number;
        for (char c : line) 
        {
            if (c == '1') number.push_back(1);
            if (c == '0') number.push_back(0);
        }
        list.push_back(number);
    }

    return list;
}


int bin2int(binary number)
{
    int value = 0;
    for (int i = 0; i < number.size(); i++)
    {
        value += number[number.size()-1-i] * std::pow(2, i);
    }
    return value;
}


long power_consumption(binaryList numbers)
{
    binary gamma, epsilon;
    for (int i = 0; i < numbers[0].size(); i++)
    {
        int acc = 0;
        for (binary number : numbers)
        {
            acc += number[i];
        }

        2 * acc > numbers.size() ? gamma.push_back(1) : gamma.push_back(0);
        2 * acc > numbers.size() ? epsilon.push_back(0) : epsilon.push_back(1);
    }
    return bin2int(gamma) * bin2int(epsilon);
}


binary bit_filter(binaryList& numbers, bool filterOnMostCommon=true)
{
    int bit_length = numbers[0].size();
    binaryList candidates = numbers;

    for (int i = 0; i < bit_length && 1 < candidates.size(); i++)
    {
        binaryList ones, zeros;

        for (binary number : candidates)
            number[i] ? ones.push_back(number) : zeros.push_back(number);

        if (filterOnMostCommon)
            candidates = ones.size() >= zeros.size() ? ones : zeros;
        else
            candidates = zeros.size() <= ones.size() ? zeros : ones;
    }
    return candidates[0];
}


long life_support_rating(binaryList numbers)
{
    binary ogr = bit_filter(numbers, true);
    binary csr = bit_filter(numbers, false);
    return bin2int(ogr) * bin2int(csr);
}


int main()
{
    binaryList number_list = read_input();
    std::cout << "Part 1. Power Consumption: " << power_consumption(number_list) << std::endl;
    std::cout << "Part 2. Life Support Rating: " << life_support_rating(number_list) << std::endl;
    return 0;
}